### DISCUSSION TITLE
scheduler vs nanny

### DISCUSSION THREAD

**Discussion by YarShev** [dask/dask#7157] [Q&A] [answered]
Hi there,

I am curious on differences between scheduler process and nanny process. Please, give me short overview of it or point me to some documentation on it.

Thanks in advance!

**Comment by jrbourbeau** [comment]
An overall overview is at https://distributed.dask.org/en/latest/#architecture. See https://distributed.dask.org/en/latest/worker.html#nanny for a description of nanny processes.

**Reply by YarShev** [reply]
@jrbourbeau , thank you! I have read these resources but I don't understand some details yet. Can you answer to these questions?
1. Scheduler process is process where `Client()` is called from?
2. If `1` clause is not correct, who creates scheduler process?
3. Who creates nanny process?
4. Nanny process and scheduler process are different processes, aren't?

**Comment by quasiben** [comment, marked as answer]
When using a single machine and `Client()` the scheduler will setup a [single process](https://docs.dask.org/en/latest/setup/single-distributed.html#client) for both client and scheduler.  This is not always true -- users can start the scheduler manually with `dask-scheduler` or with a variety of other deployment tools.  There are no nanny processes for the scheduler.  [The nanny](https://docs.dask.org/en/latest/setup/cli.html?highlight=nanny#nanny-processes) only monitors the workers.  From the docstring:


> The nanny spins up Worker processes, watches then, and kills or restarts
    them as necessary. It is necessary if you want to use the
    ``Client.restart`` method, or to restart the worker automatically if
    it gets to the terminate fraction of its memory limit.

**Reply by YarShev** [reply]
@quasiben , we are almost by the goal :) One more question. The scheduler spins up nanny processes and the latters spin up worker processes themselves. Right?

**Reply by quasiben** [reply]
No, the scheduler is only responsible for itself.  When a worker is create: dask-worker, LocalCluster, Other Cluster Manager (kubecluster, dask_cloudprovider, etc) a nanny process is also created.  In some cases, it is possible to start the worker without the nanny (dask-worker --no-nanny).

**Reply by YarShev** [reply]
Oh, I see, the nanny is a just small separate process tracking a respective worker process. Thanks a lot, @quasiben !
